src/components/SidePanels/TabPanels/Tabs.tsx

```diff
@@ -0,0 +1,178 @@
+import React, { useEffect, useRef, useState } from 'react';
+import { IconOutlinerDrag } from '../../../assets/icons';
+import { useDrag } from '../../../hooks/useDrag';
+import { Block, Elem } from '../../../utils/bem';
+import { DEFAULT_PANEL_MAX_HEIGHT } from '../constants';
+import './Tabs.styl';
+import { BaseProps, DroppableSide, TabProps } from './types';
+import { determineDroppableArea, determineLeftOrRight } from './utils';
+
+const Tab = (props: TabProps) => {
+  const { rootRef, tabTitle: tabText, tabIndex, panelKey, viewLength, children, active, panelWidth, transferTab, createNewPanel, setActiveTab } = props;
+  const [dragOverSide, setDragOverSide] = useState<DroppableSide | undefined>();
+  const tabRef = useRef<HTMLDivElement>();
+  const ghostTabRef = useRef<HTMLDivElement>();
+  const dragging = useRef(false);
+  const location = useRef({ panelKey, tabIndex });
+
+  useEffect(() => { location.current = { panelKey, tabIndex }; }, [panelKey, tabIndex]);
```

Comment:
Suggesting the same here, we don't need the useEffect to assign a ref for purposes of keeping a non stale reference to values in event handlers/callback closures.

```suggestion
  location.current = { panelKey, tabIndex };
```